### PR TITLE
Github Actions for automatic linting/testing/publishing

### LINKED ISSUE #22: Configure GitHub actions for automatic linting/testing/publishing
https://docs.github.com/en/actions/quickstart

https://beta.ruff.rs/docs/usage/#github-action

https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-python

### PR DESCRIPTION
**Pull request by Mustaballer**
## Overview

- Resolves #22 
- Using Github Actions script that reformats code with python black, and runs linters and pytests whenever you push to any branch

## Unit Tests Created

- n/a

## Steps to QA

- See Github Actions run history on my repo: https://github.com/Mustaballer/puterbot/actions

### CODE REVIEW

**Comment by 0dm** [comment]
It looks like the pytest step is failing, 
```python
==================================== ERRORS ====================================
________________ ERROR collecting tests/puterbot/test_events.py ________________
ImportError while importing test module '/home/runner/work/puterbot/puterbot/tests/puterbot/test_events.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.10.11/x64/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/puterbot/test_events.py:9: in <module>
    from puterbot.models import ActionEvent, WindowEvent
puterbot/models.py:4: in <module>
    from pynput import keyboard
/opt/hostedtoolcache/Python/3.10.11/x64/lib/python3.10/site-packages/pynput/__init__.py:40: in <module>
    from . import keyboard
/opt/hostedtoolcache/Python/3.10.11/x64/lib/python3.10/site-packages/pynput/keyboard/__init__.py:31: in <module>
    backend = backend(__name__)
/opt/hostedtoolcache/Python/3.10.11/x64/lib/python3.10/site-packages/pynput/_util/__init__.py:78: in backend
    raise ImportError('this platform is not supported: {}'.format(
E   ImportError: this platform is not supported: ('failed to acquire X connection: Bad display name ""', DisplayNameError(''))
E   
E   Try one of the following resolutions:
E   
E    * Please make sure that you have an X server running, and that the DISPLAY environment variable is set correctly
=========================== short test summary info ============================
ERROR tests/puterbot/test_events.py
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.38s ===============================```

**Comment by Mustaballer** [comment]
> DISPLAY=":0" python3 filename

It seems like setting DISPLAY=:0 directly in the GitHub Actions workflow file will not work as expected because GitHub Actions runs in a headless environment without a graphical display. I will try to find another workaround.

**Comment by Mustaballer** [comment]
@abrichr The linter is catching on a lot of stylistic errors. Do you want me to address them in this PR?

**Comment by abrichr** [comment]
@Mustaballer that would be great!

Can you also please look into publishing to PyPi? Or do you think that would be better as a separate PR? Ultimately we want users to be able to run `pip install openadapt`

**Comment by Mustaballer** [comment]
@abrichr Yeah I would like to work on the publishing with PyPi in a seperate PR, I did some research and I think I can start with publishing manually with Test PyPi(a separate instance of the Python Package Index that allows you to try distribution tools and processes without affecting the real index). And then once i get that configured I can then write a new Github Actions script that would trigger when you create a new release etc using the actual PyPi. What do you think?

**Comment by dianzrong** [comment]
@Mustaballer Do you think we also want a type checker?

**Comment by abrichr** [comment]
@Mustaballer is this ready for re-review?

**Comment by Mustaballer** [comment]
@abrichr it is ready for review! I  haven't resolved all of the lint style checks (there's quite a lot) but please review the main.yml

File: alembic/versions/b206c80f7640_add_recording_inputevent_screenshot_.py [hunk 1]
```diff
@@ -10,75 +10,99 @@
...
     # ### commands auto generated by Alembic - please adjust! ###
-    op.create_table('recording',
-    sa.Column('id', sa.Integer(), nullable=False),
-    sa.Column('timestamp', sa.Integer(), nullable=True),
-    sa.Column('monitor_width', sa.Integer(), nullable=True),
-    sa.Column('monitor_height', sa.Integer(), nullable=True),
-    sa.Column('double_click_interval_seconds', sa.Numeric(asdecimal=False), nullable=True),
-    sa.Column('double_click_distance_pixels', sa.Numeric(asdecimal=False), nullable=True),
-    sa.Column('platform', sa.String(), nullable=True),
-    sa.PrimaryKeyConstraint('id', name=op.f('pk_recording'))
```
Comment (abrichr) [inline alembic/versions/b206c80f7640_add_recording_inputevent_screenshot_.py:29]: Surprising that alembic autogenerates incorrectly formatted migration scripts. Please feel free to create a new issue on their github and link back to here: https://github.com/alembic/alembic/issues 🙏

**Review by abrichr — APPROVED** [review]
Thanks @Mustaballer ! Looks good, what else is next?

File: openadapt/strategies/__init__.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-from openadapt.strategies.base import BaseReplayStrategy
-from openadapt.strategies.naive import NaiveReplayStrategy
-from openadapt.strategies.demo import DemoReplayStrategy
+# from openadapt.strategies.base import BaseReplayStrategy
+# from openadapt.strategies.naive import NaiveReplayStrategy
+# from openadapt.strategies.demo import DemoReplayStrategy
```
Comment (abrichr) [inline openadapt/strategies/__init__.py:3]: This seems incorrect to me, I believe we need these here. @Mustaballer can you please confirm by running a replay strategy?

File: tests/openadapt/test_events.py [hunk 3]
```diff
@@ -202,58 +217,52 @@ def make_doubleclick_event(
...
-        override_double_click_interval_seconds(
-            OVERRIDE_DOUBLE_CLICK_INTERVAL_SECONDS
-        )
+        override_double_click_interval_seconds(OVERRIDE_DOUBLE_CLICK_INTERVAL_SECONDS)
     double_click_interval_seconds = get_double_click_interval_seconds()
     dt_short = double_click_interval_seconds / 10
     dt_long = double_click_interval_seconds * 10
     raw_events = [
         *make_click_events(dt_long, button_name="right"),
-
```
Comment (abrichr) [inline tests/openadapt/test_events.py:213]: I find these blank spaces to make the test more readable. @Mustaballer can you please add directives here to allow these lines?

File: tests/openadapt/test_events.py [hunk 4]
```diff
@@ -307,65 +324,103 @@ def test_merge_consecutive_mouse_scroll_events():
...
-        make_click_event(True, 2, get_post_children=lambda: [
-            make_move_event(2),
-        ]),
-        make_click_event(False, 3, get_post_children=lambda: [
-            make_move_event(3),
-        ]),
-    ])
+    expected_events = rows2dicts(
+        [
+            make_click_event(
```
Comment (abrichr) [inline tests/openadapt/test_events.py:365]: This looks less readable. Can we ignore these lines?

**Comment by Mustaballer** [comment]
I will create a new pull request to address the changes and remove all the formatting modifications. Additionally, I will use a different style, such as Google Python, instead of reStructuredText after confirming with you.

**Comment by Mustaballer** [comment]
@abrichr I can close this PR and add the github action code to a new branch off the latest main.

**Pull request closed without merging**